src/litdata/utilities/dataset_utilities.py

```diff
@@ -91,8 +92,41 @@ def _should_replace_path(path: Optional[str]) -> bool:
     return path.startswith("/teamspace/datasets/") or path.startswith("/teamspace/s3_connections/")
 
 
+def _read_last_updated_timestamp(input_dir: Optional[Dir]) -> str:
+    """Read last updated timestamp from index.json file."""
+    last_updation_timestamp = ""
+    index_json_content = None
+    assert isinstance(input_dir, Dir)
+
+    if input_dir.path is not None and os.path.exists(os.path.join(input_dir.path, _INDEX_FILENAME)):
+        # read index.json file and read last_updation_timestamp
+        index_json_content = load_index_file(input_dir.path)
+    elif input_dir.url is not None:
+        assert input_dir.url is not None
+        # download index.json file and read last_updation_timestamp
+        with tempfile.TemporaryDirectory() as tmp_directory:
+            temp_index_filepath = os.path.join(tmp_directory, _INDEX_FILENAME)
+            downloader = get_downloader_cls(input_dir.url, tmp_directory, [])
+            downloader.download_file(os.path.join(input_dir.url, _INDEX_FILENAME), temp_index_filepath)
+
+            index_json_content = load_index_file(tmp_directory)
+
+    if index_json_content is not None and "updated_at" in index_json_content:
+        last_updation_timestamp = index_json_content["updated_at"]
+
+    return last_updation_timestamp
+
+
 def _try_create_cache_dir(input_dir: Optional[str]) -> Optional[str]:
-    hash_object = hashlib.md5((input_dir or "").encode())  # noqa: S324
+    # input_dir = input_dir.path if input_dir.path else input_dir.url
+    resolved_input_dir = _resolve_dir(input_dir)
+    last_updation_timestamp = _read_last_updated_timestamp(resolved_input_dir)
+
+    if last_updation_timestamp == "":
+        # for backward compatibility, use the input_dir for hashing (if no timestamp is found)
+        last_updation_timestamp = input_dir if input_dir else ""
```

Comment:
```suggestion
        last_updation_timestamp = input_dir if input_dir else ""
```

You are hashing the entire input_dir including the `cache_dir` which can be changed by the user. So you are effectively creating non-deterministic hashes.

Reply:
but, this is just for backward compatibility.

The original code was: 
```python
hash_object = hashlib.md5((input_dir or "").encode())
```

There's already `input_dir or ""`. So, If last_timestamp is empty, just do as before.

Am I missing something?